Enumerables project

denisdiaconu opened the pull request:
In this project, we recreated some of the most famous Enumerable methods using ruby. Those methods are:
- #my_each
- #my_each_with_index
- #my_select
- #my_all?
- #my_any?
- #my_none?
- #my_count
- #my_map
- #my_inject

denisdiaconu committed:
Create .rubocop.yml

denisdiaconu committed:
Create linters.yml

denisdiaconu committed:
Create tests.yml

denisdiaconu committed:
finished Enumerables

denisdiaconu committed:
fixed linters

Ghiftee committed:
add readme

Ghiftee committed:
update readme

Ghiftee committed:
made code changes

denisdiaconu committed:
Update README.md

Georjane reviewed (approved):
## Status: Changes Required 🛑 🔧 
### Hi there Team ✋🏽 
Nice work on the enumerable methods but most of the test cases are not yet passing. You have to consider cases for when a block is given or not when an array, hash, or range is passed as an argument too. You have the following changes to make:

- [x] enumerables #my_each returns an Enumerator if no block is given
- [x] enumerables #my_each when a block is given when self is a range returns the range itself after calling the given block once for each element in self
- [x] enumerables #my_each when a block is given when self is a range behaves like #each calls the given block once for each element in self
- [x] enumerables #my_each when a block is given when self is a hash behaves like #each calls the given block once for each element in self 
- [x] enumerables #my_each_with_index returns an enumerator if no block is given
- [x] enumerables #my_each_with_index when a block is given when self is a range returns the range itself after calling the given block once for each element in self
- [x] enumerables #my_select returns an array containing all elements of enum for which the given block returns a true value
- [x] enumerables #my_select returns an array containing all elements of enum for which the given block returns a true value::range
- [x] enumerables #my_all? returns true if the block never returns false or nil
- [x] enumerables #my_all? returns false if the block returns false or nil
- [x] enumerables #my_none? when a class is passed as an argument returns true if none of the collection is a member of such class::Numeric
- [x] enumerables #my_none? when a Regex is passed as an argument returns true if none of the collection matches the Regex
- [x] enumerables #my_map executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.

Please make the following changes and submit them for another review.
_Feel free to reach out to me on Slack if you have any questions, Witah Georjane_ 🤝 
Good luck on your project and Happy Coding 💻 

**Witah Georjane
Code Reviewer at Microverse Inc
[portfolio](https://witahgeorjane.tech/) | [linkedIn](https://www.linkedin.com/in/witah-georjane/)**

denisdiaconu committed:
Fixed code review req

denisdiaconu commented:
Thank you for the code review!

fixed:
- Enumerables #my_each  - fixed all the tasks
- Enumerables #my_each_with_index  - fixed all the tasks
- Enumerables #my_select  - fixed all the tasks
- Enumerables #my_all?  - fixed all the tasks
- Enumerables #my_none?  - fixed all the tasks
- Enumerables #my_map  - fixed all the tasks

ebeagusamuel reviewed (changes requested):
# Changes Required :recycle: 
### Hey there :wave: 
###### You still have a lot of test cases failing. I would have listed them all but they are a  lot( 65 out of 79 cases are failing). I would suggest you take the time to go over how the original enumerable methods work. This will help you better understand how to mimic that behavior. It is expected that your custom methods behave exactly like that of the original enumerable method. Also, the naming of some of your method are not correct, for instance instead of `my_any?` you have `my_any`. Take into account also the suggestion made by the previous reviewer.
##### When you are done you can request another review.
slack(@ebeagusamuel)

mke2111 reviewed (changes requested):
# **🔴 STATUS: CHANGES REQUIRED 🔴 ⚠️** 

Hey @denisdiaconu & @Ghiftee, The methods worked just fine except for a few. I'll breakdown everything for you.

### REQUIRED CHANGES: 

- [x] Please make all the changes requested by the previous reviewer. Thats all you need for now.

In case you need help or don't understand a few things, reach out to the reviewers.

Happy Coding.

Reach me on Slack: **Roy Mukuye**

denisdiaconu committed:
fixed errors

denisdiaconu committed:
fixed errors

denisdiaconu commented:
Thank you for the review!

Fixed:
- Naming of the method
- fixed all the errors about ranges
-  fixed all the errors about no block given
- fixed all the errors 

Thanks for the review!

denisdiaconu committed:
fixed error

denisdiaconu committed:
fixed error

GzimAsani reviewed (changes requested):
###  Status: Changes Required :recycle:
**Hello team** very nice job implementing all the requirements, however your still have 22 failures in your code, kindly see them below:
- [x] #my_each when a block is given when self is a hash behaves like #each calls the given block once for each element in self
- [x] enumerables #my_each_with_index when a block is given when self is an array returns the array itself after calling the given block once for each element in self
- [x] enumerables #my_each_with_index when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self
- [x] enumerables #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric
- [x] enumerables #my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex
- [x] enumerables #my_any? when no block or argument is given returns false if at least one of the collection is not true
- [x] enumerables #my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric
- [x] enumerables #my_any? when a Regex is passed as an argument returns true if any of the collection matches the Regex
- [x] enumerables #my_none? returns false if the block ever returns true for all elements::range
- [x] enumerables #my_none? when no block or argument is given returns false only if one of the collection members is true
- [x] enumerables #my_none? when a Regex is passed as an argument returns false if any of the collection matches the Regex
- [x] enumerables #my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
- [x] enumerables #my_none? when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern
- [x] enumerables #my_count returns the number of items in enum through enumeration
- [x] enumerables #my_count returns the number of items in enum through enumeration::range
- [x] enumerables #my_count counts the number of items in enum that are equal to item if an argument is given
- [x] enumerables #my_map returns an Enumerator if no block is given
- [x]  enumerables #my_inject raises a "LocalJumpError" when no block or argument is given
- [x] enumerables #my_inject when a block is given with an initial value combines all elements of enum by applying a binary operation, specified by a block
- [x] enumerables #my_inject when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method:range
- [x] enumerables #my_inject when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method
- [x]  enumerables #my_inject when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method::range

### **Please make the required changes and submit another review request when you are ready.**

<hr>
### if you have any question kindly reach me on slack @GzimAsani

### HAPPY CODING :computer:

denisdiaconu committed:
fixed

denisdiaconu commented:
Thanks for the code review!
We give our best and we tried to fix everything!

GzimAsani reviewed (changes requested):
###  Status: Changes Required :recycle:
**Hello team** I know you are trying hard to complete this project, and I am sorry but you still have to work on some issues, kindly see below:
- [x] You are still having some issues with my_all?, my_none?, my_any?, kindly visit https://ruby-doc.org/core-2.6.5/Enumerable.html this page as a refresher for the requirements that you have to satisfy and try to refactor your code

denisdiaconu committed:
fixed

denisdiaconu committed:
fixed

denisdiaconu commented:
thank you for the code review!

fixed all the problems with:
 with my_all?
 my_none?
 my_any?

mke2111 reviewed (changes requested):
# **STATUS: CHANGES REQUIRED 🔴** 

### Hey Team,
Thank you for tirelessly working towards getting this project approved. We are almost there. To get there completely, we have only 6 test cases failing.

### REQUIRED CHANGES:

- [ ] ` my_each ` when a block is given when self is a hash behaves like the `each` and calls the given block once for each element in self.  For a given block, this is the result
```
 expected: "[:a, 1][:b, 2][:c, 3][:d, 4][:e, 5]"
            got: "a1b2c3d4e5"
```

- [x] ` my_each_with_index ` when a block is given when self is a hash behaves like ` each_with_index ` calls the given block once for each element in self. Right now, your enumerable doesn't behave like the ruby enumerable `each_with_index`.

- [ ]  ` my_any? ` should return false if the block ever returns a value false or nil::range when a range is passed to it. Kindly look into it and compare it to the ruby enumerable of `any?`

- [ ]  ` my_none? ` should return false if the block ever returns a value true for all elements::range when a range is passed to it. Kindly look into it and compare it to the ruby enumerable of `none?`

- [ ] ` my_count ` counts the number of items in enum that are equal to the item if an argument is given. Here is the result for a given block.
```
array.my_count(LOWEST_VALUE) should be equal to array.count(LOWEST_VALUE)

expected: 13   ---------with array.count(LOWEST_VALUE)
            got: 9   ---------with array.my_count(LOWEST_VALUE)

```

- [ ] ` my_map` should return an Enumerator if no block is given.

### Make the requested changes and your project will be gladly approved. 🤩 

### Cheers! 🍹 

Reach me on slack for a chitchat 🆘  **Roy Mukuye**

denisdiaconu committed:
fixed errors

denisdiaconu commented:
Thank you for the review!
I fix all the problems.

enaburekhan reviewed (changes requested):
# Status: Request changes:full_moon:
**Hi team!
I must sincerely commend your effort in working tirelessly to ensure this project gets approved as soon as possible:clap::clap:. However, there are still a few changes we need to make to get all the tests to pass and the project approved:muscle:.**

- [ ] Currently, we are having 48 errors out of 79. It will be time-consuming to list all of them here. so here is a link to the screen-shot given the summary: https://imgur.com/a/YJzKPpA 

The previous Tse confirmed we just had six errors to fix to get the project approved. So, what I could figure out caused this avalanche of errors could be that you made some changes to the #my_each. Recall that a lot of other methods depend on it.  
Kindly check your code and fix these issues.  
When you are done, request another review

Kindly drop your questions:point_down: if you have any and ping me on slack @ Eric Enaburekhan to respond.
Happy Coding :blush:
Eric Enaburehan

denisdiaconu committed:
fixed

denisdiaconu committed:
fix

denisdiaconu commented:
Thank you for the review!
redo all

denisdiaconu committed:
fixes

denisdiaconu commented:
Thank you for the review!
fixed last problems

enaburekhan commented on enumerable.rb:
```diff
@@ -0,0 +1,149 @@
+    true
+  end
+
+  def my_any?(block = nil)
+    arr = to_a
+    return false if (arr - [nil, false]) == []
+
+    my_each do |re|
+      if block_given?
+        return true if yield re
+      elsif block.instance_of?(Regexp)
+        return true if block =~ re
+      elsif block.instance_of?(Class)
+        return true if [re.class, re.class.superclass].include?(block)
+      elsif !block.nil?
+        return true if !x.nil? && x != false
+      else
+        return true unless re
+      end
+    end
+    return true if block.nil? && !block_given?
+
+    false
+  end
```
#my_any? when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern

NameError:
       undefined local variable or method `x' for ["cat", "door", "rod", "blade"]:Array
:point_up: error we are getting. Kindly refactor the code to ensure Regex is captured properly.

enaburekhan reviewed (changes requested):
# Status: Changes requested:full_moon:
**Congratulations so far on this milestone:clap::clap:. We are almost there. Just a few changes to make and we will be done and happy:blush:.** 
We just have two tests failing. So, let's make them pass and get this project approved:muscle:.

- [ ] #my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

- [ ] #my_any? when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern

Both tests are failing on the same error. as already highlighted below in the pull request, refactor the #my_any? to ensure Regex is captured.

Make the two above required changes and when you are done, request for another review.
Kindly drop your questions:point_down: and ping me on slack @ Eric Enaburekhan to respond.

Happy Coding:blush:
Eric Enaburekhan

denisdiaconu committed:
fix last errors

denisdiaconu commented:
Thank you for the code rev!

fixed my_any

mahmoud717 reviewed (changes requested):
### Changes required ♻️ 

**Hello there**

Great job with the project so far it looks amazing, but there are still some changes to be made:
 
---
### Requested 🔴 

- [ ] expect(%w[dog door rod blade].my_any?('cat')).to be %w[dog door rod blade].any?('cat')
     
       expected false
       got          true

my_any is still failing, it should return false if none of the elements is true.

---

please make these changes and request another code review.

if you have any questions don't hesitate to contact me on slack **MahmoudMohammad**

denisdiaconu committed:
fixed

denisdiaconu commented:
Thank you for the code rev!
fixed my_any?

mke2111 reviewed (approved):
# **🎉 STATUS: APPROVED 🎉** 

### Hey @Ghiftee  & @denisdiaconu ,
 **Congratulations 🥳 🥂 . The project meets all the requirements.**

**You may now merge to the develop branch and submit a project completion form. 👍🏽** 

### Cheers!! 🍹 

Reach me on slack for a chitchat 🆘  **Roy Mukuye**
